janus_core/geom_opt.py

```diff
@@ -0,0 +1,94 @@
+"""Geometry optimization."""
+
+from pathlib import Path
+from typing import Any, Optional
+
+from ase import Atoms
+from ase.io import read, write
+
+try:
+    from ase.filters import FrechetCellFilter as DefaultFilter
+except ImportError:
+    from ase.constraints import ExpCellFilter as DefaultFilter
+
+from ase.optimize import LBFGS
+
+
+def optimize(
+    atoms: Atoms,
+    fmax: float = 0.1,
+    dyn_kwargs: Optional[dict[str, Any]] = None,
+    filter_func: Optional[callable] = DefaultFilter,
+    filter_kwargs: Optional[dict[str, Any]] = None,
+    optimizer: callable = LBFGS,
+    opt_kwargs: Optional[dict[str, Any]] = None,
+    struct_kwargs: Optional[dict[str, Any]] = None,
+    traj_kwargs: Optional[dict[str, Any]] = None,
+) -> Atoms:
+    """Optimize geometry of input structure.
+
+    Parameters
+    ----------
+    atoms : Atoms
+        Atoms object to optimize geometry for.
+    fmax : float
+        Set force convergence criteria for optimizer in units eV/Å.
+    dyn_kwargs : Optional[dict[str, Any]]
+        kwargs to pass to dyn.run. Default is None.
+    filter_func : Optional[callable]
+        Apply constraints to atoms through ASE filter function.
+        Default is `FrechetCellFilter`.
```

Comment:
Not strictly... if you don't have `Frechet` it is `ExpCell`, should we assume this module should fail if the `import` of `FrechetFilter` fails?

Reply:
We can't expect the module to fail if importing `FrechetCellFilter` fails, as for ASE 3.22.1 it didn't exist, hence the try...except on importing, but I intended it as the default in the sense that it must be overruled, whether by the user or the import error. Can docstrings cause the module to fail?

Is there a standard way of writing an import-dependent default? Should I just use `DefaultFilter`?

Reply:
Docstrings don't cause failures unless you use `doctest` (which is a nice thing to do and we could make an issue about that as general recommendations for simple routines), but this is to be read by humans too could just go with
```suggestion
        Default is `FrechetCellFilter` if `ASE > 3.22.1` otherwise `ExpCellFilter`.
```

Reply:
Isn't that also not strictly true, as there will be a specific (but probably not very helpful to reference) commit where `FrechetCellFilter`, and the filters module as a whole, was added to ASE at some point after there were no new tags?

Reply:
`if available` then? 

That's why I wouldn't recommend referencing a commit, rather a release. 

PyPI doesn't allow multiple releases with the same name, so that release should have it and if you want to install a compatible version this you can do `pip install ASE>=3.22.1` hence my notation.

Reply:
yap I will not hold my horses...